### PR TITLE
Adds utility class for CDI representation

### PR DESCRIPTION
**Pull request by balazsracz**
This utility class adds a bunch of helper functions for dealing with SXML tree representing a CDI:
- allocation and free of c++ structures attached to XML nodes
- primitive queries
- the layout of the memory config variables in the memory space.

### CODE REVIEW

**Comment by balazsracz** [comment]
note that this PR can (and will) be merged into master after the bracz-datagram-multiple is merged.

File: src/openlcb/CDIUtils.hxx [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+        }
+    };
+
+    /// Smart pointer class for holding XML documents.
+    typedef std::unique_ptr<XMLDoc, XMLDocDeleter> xmldoc_ptr_t;
+
+    /// Helper class for deleting sxml strings.
+    struct SXmlStringDeleter
+    {
+        void operator()(const SXML_CHAR *data)
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: add Doxygen

File: src/openlcb/CDIUtils.hxx [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+namespace openlcb
+{
+
+class CDIUtils
+{
+public:
+    /// Helper class for unique_ptr to delete an XML document correctly.
+    struct XMLDocDeleter
+    {
+        void operator()(XMLDoc *doc)
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: add Doxygen

File: src/openlcb/CDIUtils.hxx [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+    /// Helper class for deleting sxml strings.
+    struct SXmlStringDeleter
+    {
+        void operator()(const SXML_CHAR *data)
+        {
+            free(const_cast<SXML_CHAR *>(data));
+        }
+    };
+    typedef std::unique_ptr<const SXML_CHAR, SXmlStringDeleter> xmlstring_t;
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: add Doxygen

File: src/openlcb/CDIUtils.hxx [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+        }
+        return nullptr;
+    }
+
+    /// Finds the name of a CDI element.
+    ///
+    /// @param node is a CDI element (segment, group, or config element).
+    /// @return finds a child tag <name> and returns its contents. Returns
+    /// empty string if no <name> was found.
+    static string find_node_name(const XMLNode *node, const char *def)
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: Missing @param def doxygen

File: src/openlcb/CDIUtils.hxx [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        /// parent group element. This is the sum of size values of the
+        /// preceding elements within the given group. Inside a repeated group
+        /// these offsets are counted from the current repetition start offset.
+        int offset_from_parent = 0;
+        /// Total number of bytes that this element occupies. This includes all
+        /// repetitions for a repeated group.
+        int size = 0;
+    };
+
+    static int get_numeric_attribute(
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: add Doxygen

File: src/openlcb/CDIUtils.hxx [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+        if ((!attr_value) || (attr_value[0] == 0))
+        {
+            return def;
+        }
+        return atoi(attr_value);
+    }
+
+    /// Allocates all userinfo structures within a segment and performs the
+    /// offset layout algorithm.
+    static void layout_segment(XMLNode *segment)
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: @param segment missing

File: src/openlcb/CDIUtils.hxx [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+            {
+                // end of iteration
+                break;
+            }
+        }
+    }
+
+    /// @return the number of replicas a group has, if it is a repeated group,
+    /// or 1 if it is a non-repeated group.
+    static int get_replication(const XMLNode *group)
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: @param group missing

File: src/openlcb/CDIUtils.hxx [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+        /// Element in the XML where we are. This is a segment or a group node.
+        const XMLNode *node_;
+        /// address in the current space of the beginning of the current
+        /// node. For segments, this is the origin; for groups this is the
+        /// address of the virtual zero-length data element at the beginning of
+        /// the group. For repeated groups this is meaningful only with a given
+        /// repetition of a group.
+        unsigned address_;
+
+        CDINodeRep()
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: add Doxygen.

File: src/openlcb/CDIUtils.hxx [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+
+        /// Resets the current representation. Use any constructor argument
+        /// set.
+        template <typename... Args> void reset(Args &&... args)
+        {
+            new (this) CDINodeRep(std::forward<Args>(args)...);
+        }
+
+        /// Gets the absolute address of a given child in the current segment.
+        unsigned get_child_address(const XMLNode *child) const
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: @param child missing

File: src/openlcb/CDIUtils.hxx [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+            HASSERT(child->father == node_);
+            NodeInfo *info;
+            get_userinfo(&info, child);
+            HASSERT(info);
+            return address_ + info->offset_from_parent;
+        }
+    };
+
+private:
+    // Static class; never instantiated.
```
Comment (bakerstu) [inline src/openlcb/CDIUtils.hxx]: add '/' to make it Doxygen.

Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 9 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 8 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 7 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 6 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 5 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 4 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 3 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 1 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

File: src/openlcb/CDIUtils.hxx (hunk 2 above)
Comment (balazsracz) [reply src/openlcb/CDIUtils.hxx]: done

**Pull request merged**